agent/util_other.go

```diff
@@ -3,15 +3,67 @@
 package agent
 
 import (
+	"fmt"
 	"os"
 	"os/exec"
+	"os/signal"
+	"log"
 )
 
-// ExecScript returns a command to execute a script
+// ExecScript returns a command to execute a script through a shell.
 func ExecScript(script string) (*exec.Cmd, error) {
 	shell := "/bin/sh"
 	if other := os.Getenv("SHELL"); other != "" {
 		shell = other
 	}
 	return exec.Command(shell, "-c", script), nil
 }
+
+// ExecSubprocess returns a command to execute a subprocess directly.
+func ExecSubprocess(args []string) (*exec.Cmd, error) {
+	if len(args) == 0 {
+		return nil, fmt.Errorf("need an executable to run")
+	}
+
+	return exec.Command(args[0], args[1:]...), nil
+}
+
+// StartSubprocess starts the given command and sets up signal relaying.
+func StartSubprocess(cmd *exec.Cmd, echoSignals bool, logger *log.Logger) error {
+	if err := cmd.Start(); err != nil {
+		return err
+	}
+
+	// If enabled, start a goroutine to relay signals to the subprocess and
+	// another to watch for its shutdown.
+	if echoSignals {
+		signalCh := make(chan os.Signal, 4)
```

Comment:
How did you come up with 4 as the buffer size here? 

Asking because the documentation on signal.Notify states that `Package signal will not block sending to c: the caller must ensure that c has sufficient buffer space to keep up with the expected signal rate`. 4 seemed like an oddly specific choice, e.g what If I sent the subprocess a bunch of SIGHUPs in quick succession?

Reply:
I took the same value we use in listening for signals on the agent: https://github.com/hashicorp/consul/blob/master/command/agent.go#L366

Reply:
That begs the question of why its set to 4 there as well. 

I would set this buffer size to something like 10. Not that we can hypothesize anything about the incoming signal rate in how people would use this, but atleast 10 is conservatively large enough. If you are sending more than 10 signals at a time to the subprocess, you are probably doing something wrong. 

With leaving it as 4, there is the possibility of missing a signal because Signal.Notify won't block when the buffer is full, and thus your go routine below will never see it.